examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/CastingPlayer-JNI.cpp

```diff
@@ -0,0 +1,152 @@
+    ChipLogProgress(AppServer, "CastingPlayer-JNI::VerifyOrEstablishConnection() called with a timeout of: %ld seconds",
+                    static_cast<long>(commissioningWindowTimeoutSec));
+
+    // Convert the CastingPlayer jlong to a CastingPlayer pointer
+    jclass castingPlayerClass         = env->GetObjectClass(thiz);
+    jfieldID _cppCastingPlayerFieldId = env->GetFieldID(castingPlayerClass, "_cppCastingPlayer", "J");
+    VerifyOrReturnValue(
+        _cppCastingPlayerFieldId != nullptr, nullptr,
+        ChipLogError(AppServer, "CastingPlayer-JNI::VerifyOrEstablishConnection() Warning: _cppCastingPlayerFieldId == nullptr"));
+
+    jlong _cppCastingPlayerValue  = env->GetLongField(thiz, _cppCastingPlayerFieldId);
+    CastingPlayer * castingPlayer = reinterpret_cast<CastingPlayer *>(_cppCastingPlayerValue);
+    VerifyOrReturnValue(
+        castingPlayer != nullptr, nullptr,
+        ChipLogError(AppServer, "CastingPlayer-JNI::VerifyOrEstablishConnection() Warning: castingPlayer == nullptr"));
+
+    // Create a new Java CompletableFuture
+    jclass completableFutureClass          = env->FindClass("java/util/concurrent/CompletableFuture");
+    jmethodID completableFutureConstructor = env->GetMethodID(completableFutureClass, "<init>", "()V");
+    jobject completableFutureObj           = env->NewObject(completableFutureClass, completableFutureConstructor);
+    jobject completableFutureObjGlobalRef  = env->NewGlobalRef(completableFutureObj);
+    VerifyOrReturnValue(
+        completableFutureObjGlobalRef != nullptr, nullptr,
+        ChipLogError(AppServer,
+                     "CastingPlayer-JNI::VerifyOrEstablishConnection() Warning: completableFutureObjGlobalRef == nullptr"));
+
+    ConnectCallback callback = [completableFutureObjGlobalRef](CHIP_ERROR err, CastingPlayer * playerPtr) {
+        ChipLogProgress(AppServer, "CastingPlayer-JNI::VerifyOrEstablishConnection() ConnectCallback called");
+        if (completableFutureObjGlobalRef == nullptr)
+        {
+            // Prevents an app crash at CallBooleanMethod
+            ChipLogError(AppServer,
+                         "CastingPlayer-JNI::VerifyOrEstablishConnection() ConnectCallback, Warning: completableFutureObjGlobalRef "
+                         "== nullptr");
+        }
+        else
+        {
+            JNIEnv * env                  = JniReferences::GetInstance().GetEnvForCurrentThread();
+            jclass completableFutureClass = env->FindClass("java/util/concurrent/CompletableFuture");
+
+            if (err == CHIP_NO_ERROR)
+            {
+                ChipLogProgress(
+                    AppServer,
+                    "CastingPlayer-JNI::VerifyOrEstablishConnection() ConnectCallback, Casting Player connection successful!");
+                jmethodID completeMethod = env->GetMethodID(completableFutureClass, "complete", "(Ljava/lang/Object;)Z");
+                env->CallBooleanMethod(completableFutureObjGlobalRef, completeMethod, NULL);
```

Comment:
To avoid deadlocks, before calling the client code back on the Future, we may need to Unlock the stack (just like the API locks the stack in the beginning).

```
chip::DeviceLayer::StackUnlock unlock;
```

Reply:
Also, if we have not done so yet with the Discovery API callbacks, we will need to Unlock the stack before calling back on the CastingPlayerChangeListener handlers as well.

Reply:
Added.